sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessor.cs

```diff
@@ -198,58 +239,331 @@ public async Task StopAsync()
+
+            // Create the new partition pump and add it to the dictionary.
+
+            var partitionContext = new PartitionContext(InnerClient.EventHubName, ConsumerGroup, partitionId);
+            var checkpointManager = new CheckpointManager(partitionContext, Manager, Identifier);
+
+            try
+            {
+                var partitionProcessor = PartitionProcessorFactory(partitionContext, checkpointManager);
+                var partitionPump = new PartitionPump(InnerClient, ConsumerGroup, partitionId, partitionProcessor, Options);
+
+                await partitionPump.StartAsync().ConfigureAwait(false);
+
+                PartitionPumps[partitionId] = partitionPump;
+            }
+            catch (Exception)
+            {
+                // If partition pump creation fails, we'll try again on the next time this method is called.  This should happen
+                // on the next load balancing loop as long as this instance still owns the partition.
+                // TODO: delegate the exception handling to an Exception Callback.
+            }
+        }
+
+        /// <summary>
+        ///   Stops an owned partition pump instance in case it exists.  It is also removed from the pumps dictionary.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition the partition pump is associated with.</param>
+        /// <param name="reason">The reason why the partition processor is being closed.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task RemovePartitionPumpIfItExistsAsync(string partitionId,
+                                                              PartitionProcessorCloseReason? reason = null)
+        {
+            if (PartitionPumps.TryRemove(partitionId, out var pump))
+            {
+                try
+                {
+                    await pump.StopAsync(reason).ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    // TODO: delegate the exception handling to an Exception Callback.
+                }
+            }
+        }
+
+        /// <summary>
+        ///   Tries to claim the ownership of the specified partition.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition the ownership is associated with.</param>
+        /// <param name="completeOwnershipList">A complete list of ownership obtained from the stored service provided by the user.</param>
+        ///
+        /// <returns>The claimed ownership. <c>null</c> if the claim attempt failed.</returns>
+        ///
+        private async Task<PartitionOwnership> ClaimOwnershipAsync(string partitionId,
+                                                                   List<PartitionOwnership> completeOwnershipList)
```

Comment:
Any reason why you went with a `List` here?  The way that you're using it, an Enumerable would seem to suffice and give you better flexibility when calling.

Reply:
No particular reason. I'll update it.